SQL: Export array values through result sets

bpintea opened the pull request:
This adds to the JDBC driver the capability of exporting each arrays as
an individual result set, enabling this feature for the clients
supporting it.

bpintea committed:
SQL: Introduce core support for multivalue fields (#67883)

* Introduce core support for multivalue fields

This introduces support for multivalue fields, whose value can be
retrieved with a newly added ARRAY function. Only value extraction is of
concern here.

bpintea committed:
SQL: Add text formatting support for multivalue (#68606)

* Add text formatting support for multivalue

This adds the text formatting for multivalue doc fields, that is:
- CSV and TSV exports
- TXT formatting

bpintea committed:
Disable array test during merging

This disables one array test that was left active past merging.

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
SQL: Add xDBC and CLI support. QA CSV specs (#68966)

This adds support for the xDBC and CLI clients.
Based on those, the CsvJdbc-based QA tests are also integrated.

bpintea committed:
Merge branch 'master' into feat/sql-multivalue

bpintea committed:
Export array values through result sets

This adds to the JDBC driver the capability of exporting each arrays as
an individual result set, enabling this feature for the clients
supportig it.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcDatabaseMetaData.java:
```diff
@@ -1258,10 +1259,10 @@ public boolean generatedKeyAlwaysReturned() throws SQLException {
                     if (next instanceof EsType || next instanceof JDBCType) {
                         try {
-                            type = TypeUtils.of((JDBCType) next);
+                            type = TypeUtils.of((SQLType) next);
```
Casting an EsType to JDBCType triggers CCE, SQLType is the common type (and what TypeUtils#of takes).

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

bpintea commented:
This also addresses some points raised in https://github.com/elastic/elasticsearch/pull/68966#pullrequestreview-594024373, but without the [types enum rework](https://github.com/elastic/elasticsearch/pull/68966#discussion_r579091178) (which will be addressed in a subsequent PR, as the outcome might be debatable).

matriv commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java:
```diff
@@ -8,49 +8,60 @@
 
-    public JdbcArray(EsType type, List<?> values) {
-        this.type = type;
+    JdbcArray(JdbcConfiguration cfg, EsType baseType, List<?> values) {
```
I personally don't like that `JdbcArray `constructor needs the cfg now.
What are the actual variables of cfg needed by `memorySet()` internally?
Can you check if we can have another version of memorySet where cfg is not needed?

Can we at least have a second constructor which passed null for cfg?
I'm thinking a bit into the future `JdbcArray` can be used for setting query params to a preparedStmt and then cfg will be needed, when it's actually only for getting it as a `resultSet`.

bpintea replied:
> I personally don't like that JdbcArray constructor needs the cfg now.

Makes sense.

> Can you check if we can have another version of memorySet where cfg is not needed?

We don't. The `JdbcResultSet` constructor itself needs it, but for one thing only: to get a [calendar instance](https://github.com/elastic/elasticsearch/blob/1ee4c50217b56c0f7864ec3ed4b81c5a4d96309a/x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcResultSet.java#L71) (based on the TZ). So a `null` isn't feasible where a ResultSet needs to be built out of an array, unless we relax the `defaultCalendar` to `@Nullable` inside it.

bpintea replied:
Could refactor to have `JdbcResultSet` only take the TZ and a `JdbcArray` just provide a default (`Z`) when not available. But wondering if this change should be made now, or when needed? I'm fine with either.

matriv replied:
I'd genuinely, prefer to do that now, it's a too much to add the whole cfg.

bpintea replied:
The original intention of passing the entire cfg might have stemmed from the intention to have locales passed in the session configuration too? In any case..

> Can we at least have a second constructor which passed null for cfg?
I'm thinking a bit into the future JdbcArray can be used for setting query params to a preparedStmt

I've updated to now only take the time zone, but no null TZ constructor, since I guess this will be available / needed also for the prepared statement case? OTOH, not sure what a good default would be here - Z or default - for cases the user passes an array of millis which would maybe need to be converted internally by the driver as, for instance, timestamps?

matriv reviewed (commented):
I have an "objection" regarding the use of cfg in the ctor.

costin reviewed (approved):
LGTM.
Re: our discussion on EsType / enum refactoring I assume there's going to be a follow-up PR?

bpintea committed:
Address review feedback

- make certain members package private.
- replace passing around the entire session configuration with just the
  timezone in JdbcResultSet & co.

bpintea commented:
> Re: our discussion on EsType / enum refactoring I assume there's going to be a follow-up PR?

[Yes](https://github.com/elastic/elasticsearch/pull/69512#issuecomment-784995751), correct.

bpintea commented:
@elasticmachine run elasticsearch-ci/2 elasticsearch-ci/bwc

palesz commented on x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/JdbcArrayTests.java:
```diff
@@ -66,4 +75,90 @@ public void testArraySlicing() throws SQLException {
+        assertFalse(resultSet.next());
+    }
+
+    public void testArrayGetAsSlicedResultSet() throws SQLException {
```
Maybe add some randomness here instead of the constants.

bpintea replied:
I could another one, but I'd like to keep this one as is, as it's very simple to follow how the slicing is done and what's expected. (and with some willingness from the gods of CI, not have to run the whole suite again)

palesz reviewed (approved):
LGTM.

matriv reviewed (approved):
LGTM

bpintea commented:
@elasticmachine run elasticsearch-ci/bwc

astefan reviewed (approved):
LGTM

bpintea commented:
@elasticmachine run elasticsearch-ci/bwc

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-multivalue' into feat/sql-multivalue

bpintea committed:
Merge branch 'feat/sql-multivalue' into feat/jdbc-array-rs
